sweepbatcher/sweep_batch.go

```diff
@@ -0,0 +1,1369 @@
+		errChan:          make(chan error, 1),
+		callEnter:        make(chan struct{}),
+		callLeave:        make(chan struct{}),
+		quit:             make(chan struct{}),
+		completeChan:     bk.completeChan,
+		batchTxid:        bk.batchTxid,
+		batchPkScript:    bk.batchPkScript,
+		rbfCache:         bk.rbfCache,
+		wallet:           bk.wallet,
+		chainNotifier:    bk.chainNotifier,
+		signerClient:     bk.signerClient,
+		muSig2SignSweep:  bk.musig2SignSweep,
+		verifySchnorrSig: bk.verifySchnorrSig,
+		purger:           bk.purger,
+		store:            bk.store,
+		log:              bk.log,
+		cfg:              &cfg,
+	}
+}
+
+// addSweep tries to add a sweep to the batch. If this is the first sweep being
+// added to the batch then it also sets the primary sweep ID.
+func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
+	done, err := b.scheduleNextCall()
+	defer done()
+
+	if err != nil {
+		return false, err
+	}
+
+	// If the provided sweep is nil, we can't proceed with any checks, so
+	// we just return early.
+	if sweep == nil {
+		return false, nil
+	}
+
+	// Before we run through the acceptance checks, let's just see if this
+	// sweep is already in our batch. In that case, just update the sweep.
+	_, ok := b.sweeps[sweep.swapHash]
+	if ok {
+		b.sweeps[sweep.swapHash] = *sweep
+
+		// If this is the primary sweep, we also need to update the
+		// batch's confirmation target.
+		if b.primarySweepID == sweep.swapHash {
+			b.cfg.batchConfTarget = sweep.confTarget
+		}
+
+		return true, nil
+	}
+
+	// Since all the actions of the batch happen sequentially, we could
+	// arrive here after the batch got closed because of a spend. In this
+	// case we cannot add the sweep to this batch.
+	if b.state != Open {
+		return false, nil
+	}
+
+	// If this batch contains a single sweep that spends to a non-wallet
+	// address, or the incoming sweep is spending to non-wallet address,
+	// we cannot add this sweep to the batch.
+	for _, s := range b.sweeps {
+		if s.isExternalAddr || sweep.isExternalAddr {
+			return false, nil
+		}
+	}
+
+	// Check the timeout of the incoming sweep against the timeout of all
+	// already contained sweeps. If that difference exceeds the configured
+	// maximum we cannot add this sweep.
+	for _, s := range b.sweeps {
+		timeoutDistance :=
+			int32(math.Abs(float64(sweep.timeout - s.timeout)))
+
+		if timeoutDistance > b.cfg.maxTimeoutDistance {
+			return false, nil
+		}
+	}
+
+	// Past this point we know that a new incoming sweep passes the
+	// acceptance criteria and is now ready to be added to this batch.
+
+	// If this is the first sweep being added to the batch, make it the
+	// primary sweep.
+	if b.primarySweepID == lntypes.ZeroHash {
+		b.primarySweepID = sweep.swapHash
+		b.cfg.batchConfTarget = sweep.confTarget
+
+		// We also need to start the spend monitor for this new primary
+		// sweep.
+		err := b.monitorSpend(ctx, *sweep)
+		if err != nil {
+			return true, err
```

Comment:
False?

Reply:
Hmmm, I think you're right, semantically the sweep is "added", but in storage it is not persisted, better return `False` :+1: